cdc/kv/shared_client.go

```diff
@@ -0,0 +1,1202 @@
+	s.totalSpans.v = make(map[SubscriptionID]*requestedTable)
+	s.initMetrics()
+	return s
+}
+
+// AllocSubscriptionID gets an ID can be used in `Subscribe`.
+func (s *SharedClient) AllocSubscriptionID() SubscriptionID {
+	return SubscriptionID(subscriptionIDGen.Add(1))
+}
+
+// Subscribe the given table span.
+// NOTE: `span.TableID` must be set correctly.
+func (s *SharedClient) Subscribe(
+	subID SubscriptionID, span tablepb.Span, startTs uint64,
+	eventCh chan<- MultiplexingEvent,
+) (existSubID SubscriptionID, success bool) {
+	if span.TableID == 0 {
+		log.Panic("event feed subscribe with zero tablepb.Span.TableID",
+			zap.String("namespace", s.changefeed.Namespace),
+			zap.String("changefeed", s.changefeed.ID))
+	}
+
+	s.totalSpans.Lock()
+	if existSubID, success = s.totalSpans.m.Get(span); success {
+		s.totalSpans.Unlock()
+		success = false
+		return
+	}
+
+	rt := s.newRequestedTable(subID, span, startTs, eventCh)
+	s.totalSpans.m.ReplaceOrInsert(span, subID)
+	s.totalSpans.v[subID] = rt
+	s.totalSpans.Unlock()
+
+	s.requestRangeCh.In() <- rangeTask{span: span, requestedTable: rt}
+	log.Info("event feed subscribes table success",
+		zap.String("namespace", s.changefeed.Namespace),
+		zap.String("changefeed", s.changefeed.ID),
+		zap.Any("subscriptionID", rt.subscriptionID),
+		zap.String("span", rt.span.String()))
+	success = true
+	return
+}
+
+// Unsubscribe the given table span. All covered regions will be deregistered asynchronously.
+// NOTE: `span.TableID` must be set correctly.
+func (s *SharedClient) Unsubscribe(span tablepb.Span) (subID SubscriptionID, success bool) {
+	if span.TableID == 0 {
+		log.Panic("event feed unsubscribe with zero tablepb.Span.TableID",
+			zap.String("namespace", s.changefeed.Namespace),
+			zap.String("changefeed", s.changefeed.ID))
+	}
+
+	s.totalSpans.Lock()
+	if subID, success = s.totalSpans.m.Get(span); !success {
+		s.totalSpans.Unlock()
+		return
+	}
+
+	s.totalSpans.m.Delete(span)
+	rt := s.totalSpans.v[subID]
+	s.totalSpans.Unlock()
+
+	log.Info("event feed unsubscribes table",
+		zap.String("namespace", s.changefeed.Namespace),
+		zap.String("changefeed", s.changefeed.ID),
+		zap.Any("subscriptionID", rt.subscriptionID),
+		zap.String("span", rt.span.String()))
+
+	s.setTableStopped(rt)
+	return
+}
+
+// ResolveLock is a function. If outsider subscribers find a span resolved timestamp is
+// advanced slowly or stopped, they can try to resolve locks in the given span.
+func (s *SharedClient) ResolveLock(span tablepb.Span, maxVersion uint64) (success bool) {
+	if span.TableID == 0 {
+		log.Panic("event feed unsubscribe with zero tablepb.Span.TableID",
+			zap.String("namespace", s.changefeed.Namespace),
+			zap.String("changefeed", s.changefeed.ID))
+	}
+
+	s.totalSpans.Lock()
+	var subID SubscriptionID
+	if subID, success = s.totalSpans.m.Get(span); success {
+		rt := s.totalSpans.v[subID]
+		s.totalSpans.Unlock()
+		rt.updateStaleLocks(s, maxVersion)
```

Comment:
It is more appropriate to return `resolveLockTask slice` in `updateStaleLocks` to avoid using complex callback function (postUpdateRegionResolvedTs).